ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/State.hs

```diff
@@ -276,28 +270,21 @@ withOpenState ImmutableDBEnv { _dbHasFS = hasFS :: HasFS m h, .. } action = do
       ExitCaseSuccess (Left (UserError {}))       -> return ()
 
     shutDown :: m ()
-    shutDown =
+    shutDown = do
+      -- We need to get the most recent state because it might have changed
+      -- behind our back
+      modifyMVar _dbInternalState $ DbClosed . closedStateFromInternalState
       -- This function ('withOpenState') can be called from all threads, not
       -- necessarily the one that opened the ImmutableDB and the registry, so
       -- we must use 'unsafeReleaseAll' instead of 'releaseAll'.
-      unsafeReleaseAll _dbRegistry `finally` do
-        -- We need to get the most recent state because it might have changed
-        -- behind our back
-        st <- updateMVar _dbInternalState $ \st ->
-                (DbClosed (closedStateFromInternalState st), st)
-        closeOpenHandles hasFS st
+      unsafeReleaseAll _dbRegistry
 
     access :: HasCallStack
            => InternalState m hash h
            -> m r
     access (DbClosed _) = throwUserError _dbErr ClosedDBError
     access (DbOpen ost) = action hasFS ost
 
-closeOpenHandles :: Monad m => HasFS m h -> InternalState m hash h -> m ()
-closeOpenHandles hasFS = \case
-    DbClosed _       -> return ()
-    DbOpen openState -> closeOpenStateHandles hasFS openState
-
 -- TODO what if this fails? Use a 'ResourceRegistry'?
```

Comment:
Update comment: if one of the `hClose`s fails, it will bubble up and the bracketed `withRegistry` will close the registry and thus the remaining handles.